apps/aefate/src/aefa_fate_op.erl

```diff
@@ -1247,6 +1261,134 @@ spend(Arg0, Arg1, ES0) ->
             aefa_fate:abort({value_does_not_match_type, Other, address}, ES2)
     end.
 
+create(Arg0, Arg1, Arg2, ES0) ->
+    ?AVAILABLE_FROM(?VM_FATE_SOPHIA_2, ES0),
+    {[?FATE_CONTRACT_BYTEARRAY(Code), InitArgsTypes, Value], ES1} =
+        get_op_args([Arg0, Arg1, Arg2], ES0),
+    create_common(Code, InitArgsTypes, Value, no_gas_cap, ES1).
+
+create_common(Code, InitArgsTypes, Value, GasCap, ES0) ->
+    {ContractPK, ES1} = create_contract({new, Code}, Value, ES0),
+    ReplaceInitResult
+        = fun(ES0_) ->
+                  %% init returns unit, so we get rid of it
+                  {{tuple, {}}, ES1_} = get_op_arg({stack, 0}, ES0_),
+                  write({stack, 0}, ?FATE_CONTRACT(ContractPK), ES1_)
+          end,
+    {ok, ES2} = remote_call_common(
+           ?FATE_CONTRACT(ContractPK),
+           init,
+           InitArgsTypes,
+           ?FATE_TYPEREP({tuple, []}),
+           0,
+           GasCap,
+           unprotected,
+           ReplaceInitResult,
+           ES1),
+    {jump, 0, ES2}.
+
+clone(Arg0, Arg1, Arg2, Arg3, ES0) ->
+    ?AVAILABLE_FROM(?VM_FATE_SOPHIA_2, ES0),
+    {[?FATE_CONTRACT(CloneePK), InitArgsTypes, Value, Prot], ES1} =
+        get_op_args([Arg0, Arg1, Arg2, Arg3], ES0),
+    clone_common(CloneePK, InitArgsTypes, Value, no_gas_cap, Prot, ES1).
+
+clone_g(Arg0, Arg1, Arg2, Arg3, Arg4, ES0) ->
+    ?AVAILABLE_FROM(?VM_FATE_SOPHIA_2, ES0),
+    {[?FATE_CONTRACT(CloneePK), InitArgsTypes, Value, GasCap, Prot], ES1} =
+        get_op_args([Arg0, Arg1, Arg2, Arg3, Arg4], ES0),
+    clone_common(CloneePK, InitArgsTypes, Value, {gas_cap, GasCap}, Prot, ES1).
+
+clone_common(CloneePK, InitArgsTypes, Value, GasCap, Prot, ES0) ->
+    Protected =
+        case Prot of
+            false -> unprotected;
+            true  -> protected;
+            _     -> aefa_fate:abort({value_does_not_match_type, Prot, bool}, ES0)
+        end,
+
+    {ok, FinalCloneePK} =
+        aefa_engine_state:contract_find_final_ref(CloneePK, ES0),
+
+    {ContractPK, ES1} = create_contract({clone, FinalCloneePK}, Value, ES0),
+    ReplaceInitResult
+        = fun(ES0_) ->
+                  %% init returns unit, so we get rid of it
+                  %% if the call was protected, rewrap the result
+                  case Protected of
+                      unprotected ->
+                          {{tuple, {}}, ES1_} = get_op_arg({stack, 0}, ES0_),
+                          write({stack, 0}, ?FATE_CONTRACT(ContractPK), ES1_);
+                      protected ->
+                          case get_op_arg({stack, 0}, ES0_) of
```

Comment:
I don't remember exactly how `protected` works, but the compiler will generate code that wraps the result with a `Some`? If so, then we should never get the second (`None`) case or I am missing something?

Reply:
Ulf explained...

Reply:
I don't get the issue. It works in tests, the compiler should be adjustable to this behavior?